src/fromager/bootstrapper.py

```diff
@@ -1127,3 +1631,90 @@ def _add_to_build_order(
             # Requirement and Version instances that can't be
             # converted to JSON without help.
             json.dump(self._build_stack, f, indent=2, default=str)
+
+    def write_test_mode_report(self, work_dir: pathlib.Path) -> None:
+        """Write test mode failure report to JSON files.
+
+        Generates two JSON files:
+        - test-mode-failures.json: Detailed list of all failures
+        - test-mode-summary.json: Summary statistics with category breakdown
+        """
+        if not self.test_mode:
+            return
+
+        failures_file = work_dir / "test-mode-failures.json"
+        summary_file = work_dir / "test-mode-summary.json"
+
+        # Generate failures report
+        failures_data = {
+            "failures": [build_result.to_dict() for build_result in self.failed_builds]
+        }
+
+        with open(failures_file, "w") as f:
+            json.dump(failures_data, f, indent=2)
+        logger.info("test mode: wrote failure details to %s", failures_file)
+
+        # Generate summary report with category and exception breakdowns
+        exception_counts: dict[str, int] = {}
+        category_counts: dict[str, int] = {}
+        for build_result in self.failed_builds:
+            exception_type = build_result.exception_type or "Unknown"
+            exception_counts[exception_type] = (
+                exception_counts.get(exception_type, 0) + 1
+            )
+            category = build_result.category.value
+            category_counts[category] = category_counts.get(category, 0) + 1
+
+        summary_data = {
+            "total_packages": len(self._build_stack),
+            "total_failures": len(self.failed_builds),
+            "category_breakdown": category_counts,
+            "exception_breakdown": exception_counts,
+        }
+
+        with open(summary_file, "w") as f:
+            json.dump(summary_data, f, indent=2)
+        logger.info("test mode: wrote summary to %s", summary_file)
+
+    def finalize(self, work_dir: pathlib.Path) -> int:
+        """Finalize bootstrap and return exit code.
+
+        In test mode, writes failure reports and logs summary.
+        Returns non-zero exit code if there were failures.
+
+        Args:
+            work_dir: Directory to write reports to
+
+        Returns:
+            0 if all packages built successfully (or not in test mode)
+            1 if any packages failed in test mode
+        """
+        if not self.test_mode:
+            return 0
+
+        self.write_test_mode_report(work_dir)
+
+        if self.failed_builds:
```

Comment:
This method would have less indented code if this condition was reversed and returned 0 here.